flowauth/frontend/src/RoleDetails.jsx

```diff
@@ -0,0 +1,320 @@
+          console.log("Role err:" + err)
+          if (err.code !== 404){
+            setRole({});
+            setErrors(err.message);
+            setIsErrored(true);
+          }
+        })
+     }
+    
+  }, [])
+
+  //When Role changes, replace role.name, role.server and role.members with 
+  //the parts from the others.
+  useEffect(() => {
+      if (serverList !== []){
+        console.log("Trying to update the UI using the following role...")
+        console.log(role)
+        if (Object.keys(role).length !== 0){   //ffs, Javascript
+          console.log("Role not empty")
+          setRoleName(role.name);
+          setServer(role.server);
+          setMembers(role.users);
+          setExpiryDate(role.latest_token_expiry);
+          setMaxLifetime(String(role.longest_token_life_minutes));
+          setScopes(role.scopes)
+          setEditMode(true);
+        } else {
+          console.log("Role empty, setting defaults")
+          setRoleName("");
+          setServer(-1);
+          setMembers([]);
+          setScopes([])
+          setEditMode(false);
+      }
+    }
+  }, [role, serverList])
+
+  //Validate Rolename on change
+  useEffect(() => {
+    console.log("Name:" + name)
+    var letters = /^[A-Za-z0-9_]+$/;
+    if (name.match(letters)) {
+      setNameHelperText("");
+      setNameIsValid(true)
+    } else if (name.length === 0) {
+      setNameHelperText("Role name can not be blank.");
+      setNameIsValid(false)
+    } else {
+      setNameHelperText(
+        "Role name may only contain letters, numbers and underscores.",
+      )
+      setNameIsValid(false)
+    };
+  }, [name])
+
+  //Validate lifetime on change
+  useEffect(() => {
+    console.log("New lifetime: " + maxLifetime)
+    var numbers = /^[0-9]+$/;
+    if (maxLifetime.match(numbers)){
+      setLifetimeHelperText("");
+      setLifetimeIsValid(true)
+    } else if (maxLifetime == "") {
+      setLifetimeHelperText("Maximum lifetime cannot be blank");
+      setLifetimeIsValid(false)
+    } else {
+      setLifetimeHelperText("Lifetime must be a number")
+      setLifetimeIsValid(false)
+    }
+  }, [maxLifetime])
+
+  //Make sure entire form is valid on change for any relevant members
+  useEffect(() => {
+    const valid_vars = validation_vars.map(v => typeof(v))
+    if (valid_vars.includes("undefined")){
+      setFormIsValid(false)
+    } else if (nameIsValid && lifetimeIsValid && scopes.length > 0 && members.length > 0 && server_id >= 0){
+      console.log("Form is valid")
+      setFormIsValid(true)
+    } else {
+      setFormIsValid(false)
+    }
+  }, validation_vars)
+
+  const handleNameChange = (event) => {
+    console.log("Name change event handled")
+    setRoleName(event.target.value)
+  }
+
+  const handleLifetimeChange = (event) => {
+    console.log("Lifetime change event handled");
+    setMaxLifetime(event.target.value)
+  }
+
+  const handleServerChange = (event) => {
+    console.log("Server picker event handled");
```

Comment:
I think this may be ever so slightly too verbose.
OK during development but now I think it can safely go?!